### ISSUE TITLE
Outline: support for `//#region` fold regions

### ISSUE THREAD

**Issue by julien-c** [microsoft/vscode#51364]
in conjunction with [hierarchical support](https://github.com/Microsoft/vscode/issues/34968) in outlines.

Here's how Xcode is doing it:

<img width="485" alt="capture d ecran 2018-06-07 a 11 37 44" src="https://user-images.githubusercontent.com/326577/41110400-421f2f60-6a47-11e8-96bd-6e3b09a0a122.png">

**Comment by aeschli** [comment]
Each language support defines what document symbols it returns and these are shown in the outline.

What language do you want the feature for?

**Comment by julien-c** [comment]
Typescript, but any way of exposing a "section" symbol from a document would work, I guess.

**Comment by mjbvz** [comment]
@jrieken @aeschli I don't think this should be language specific. With folding providers and symbol providers, we already have both regions and which symbols they contain. If we want to support this feature, we should use these to construct the outline

**Comment by aeschli** [comment]
The purpose of document symbols is very spcifically to provide the symbols for the quick outline as well as the new outline. Why trying to be smart on our side and merge in the two providers with very specific purposes. This just asks for trouble.
The language provider knows much better what document symbols are meaningful to the user.

Also, the folding provider doesn't define a name that could be used as a symbol name, has a separate lifecycle...

**Comment by jrieken** [comment]
> Why trying to be smart on our side and merge in the two providers with very specific purposes. This just asks for trouble.

Well, you don't need to merge but you can simply add your own tree. Similar to the unit-test-navigator https://marketplace.visualstudio.com/items?itemName=roblourens.unittest-navigate. However, that should and easily can be done as extension. It's the combination of an API command and the registration of a document symbol provider. That will give you two separate trees and one flat view in the quick outline.

<img width="867" alt="screen shot 2018-06-22 at 12 36 19" src="https://user-images.githubusercontent.com/1794099/41856840-5983e3fa-7896-11e8-8522-0346a867c53b.png">


If you want regions/pragmas be intermixed with document symbols you need to do this per language brain/extension. 

@julien-c What is it you are looking for?

**Comment by jrieken** [comment]
Giving this back to you @mjbvz. I think this should come from language extension themselves because on then the hierarchy looks nice, e.g. not multiple root nodes per extensions/scope.

**Comment by david50407** [comment]
Is there any extension support showing Typescript region in Outline?

**Comment by fuzzy76** [comment]
Most of the discussion here is probably outdated. Regions is a cross-language thing now AFAIK, so the data is available?

**Comment by qiulang** [comment]
Hi, what is the progress of feature request? I find WebStorm also supports this nicely while all the extensions for region on marketplace look quite shabby.

<img width="488" alt="webstorm region viewer" src="https://user-images.githubusercontent.com/516243/196629907-e96b7ddd-c863-4415-9fd3-0dde79f35377.png">

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!

**Comment by fuzzy76** [comment]
Here it is then: please reconsider.

**Comment by dandeancook** [comment]
Yes please, vscode team, consider, long code files need regions to organise groups of functions/methods

**Comment by insin** [comment]
For anyone else still looking for this feature, the Region Viewer Extension is good if you use regions primarily to navigate and fold large files and you're not really bothered about having the rest of the Outline contents inside the region:

https://github.com/berabue/vscode-region-viewer

![Screen Shot 2023-07-13 at 9 41 58 am](https://github.com/microsoft/vscode/assets/226692/5ddf5453-8c7d-4fa9-99d6-bfd9f9b7b3e1)

**Comment by JoElfner** [comment]
Any news on this? This is - imho - a very important feature for usability.

**Comment by alythobani** [comment]
Hey everyone, since this isn't built into VSCode yet (and Region Viewer is great, but doesn't include language symbols + seems inactive), I put together an extension that adds this and a few other region-related tools.

[**Region Helper**](https://marketplace.visualstudio.com/items?itemName=AlyThobani.region-helper) ([GitHub](https://github.com/alythobani/vscode-region-helper))

One note: VSCode's API doesn't allow merging regions into the built-in Outline (except as a separate region-only section), so this adds a separate "Full Outline" view instead.

**Comment by jrc** [comment]
As a former Xcode user, I really miss this. With large files, I find VS Code's Outline feature to be mostly useless, as it's so noisy that it doesn't help give me a quick overview of the file. Regions would really help here.

**Comment by dbookstaber** [comment]
I'm the author of the [Outline Plus](https://github.com/dbookstaber/vscode-outline-plus) extension of [Region Helper](https://github.com/alythobani/vscode-region-helper), which I built partly because this issue closed out-of-scope. It surfaces user-defined `#region` markers in the outline view across ~50 languages and adds visibility/modifier icons (`public` / `private` / `static` / etc.). Posting once here in case it helps people who land on this thread. [Marketplace](https://marketplace.visualstudio.com/items?itemName=DavidBookstaber.outline-regions-plus). [Repo](https://github.com/dbookstaber/vscode-outline-plus).

**Comment by alythobani** [comment]
> I'm the author of the [Outline Plus](https://github.com/dbookstaber/vscode-outline-plus) extension, which I built partly because this issue closed out-of-scope. It surfaces user-defined `#region` markers in the outline view across ~50 languages and adds visibility/modifier icons (`public` / `private` / `static` / etc.). Posting once here in case it helps people who land on this thread. [Marketplace](https://marketplace.visualstudio.com/items?itemName=DavidBookstaber.outline-regions-plus). [Repo](https://github.com/dbookstaber/vscode-outline-plus).

@dbookstaber 

I normally wouldn't mind someone forking my open-source extension for their own purposes, but I would expect some more transparency than writing that you "built" it, using almost the exact same README as my own (including GIF demos I recorded myself, that even show my name in the breadcrumbs), and zero reference to the original project in your repo, Marketplace listing, or when promoting it...

Also because you didn't actually fork it (you copied it in another way?) it doesn't even show up as a fork...

This 100% comes off as plagiarism and is pretty weird to say the least.

**Comment by dbookstaber** [comment]
@alythobani fair point on attribution. I've added an Acknowledgements section to the [README](https://github.com/dbookstaber/vscode-outline-plus#acknowledgements) crediting Region Helper as the upstream and edited the comment above to call out the lineage and link your repo. The GIFs in the README now note they're from upstream.  Apologies for the earlier framing — should have done this from the start.

**Comment by alythobani** [comment]
@dbookstaber Appreciate the changes. But for proper transparency I would actually expect:

- The README to be explicit about what your fork adds/changes that didn't already exist in Region Helper
    - It looks like the only changes are modifier icons in the Full Outline view, and a couple debugging utilities. While everything else, including every core feature, was entirely built by me? If I'm wrong, an explicit list of changes would easily clarify this.
    - The "The region-navigation features, tree views and extension API originated there." line in your Acknowledgments is quite an understatement and misses a lot. Again, from what I can tell, the vast majority of your fork is just a direct copy of my extension. If you want to stop taking false credit for my work, just be clear and explicit about what your fork actually adds/changes.
        - This would also be important for users to know when choosing between our versions. Even ignoring that you initially tried to pass off the entire extension as your own (with an entirely new name, after removing all references to the original), it's pretty strange you would publish and promote a fork without making it clear how it even differs from the original.
- In general, and in your original comment: No longer stating that you "built" this extension, and being more honest and precise about what you actually did (e.g. "I made a fork of Region Helper that adds X/Y/Z")
    - Your comment above, even after your recent edit, still states you built the extension due to this issue... which is false, I built the extension due to this issue (and a few others), [as seen above](https://github.com/microsoft/vscode/issues/51364#issuecomment-2715665690). You later forked it.
    - Your comment also still says "It surfaces user-defined #region markers in the outline view across ~50 languages", which is something my extension already did. Just say what you actually added/changed, not what already existed.

Up to you if you want to make these changes or not. But they're the minimum I'd personally expect here to no longer consider this plagiarism.

**Comment by mantasu** [comment]
Just created an extension that injects regions (+comment headers) into BUILT-IN outline: https://github.com/mantasu/outline-sections

Tho I hope vscode will add support at least for folding regions
